blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala

```diff
@@ -172,5 +77,91 @@ object BlazeClient {
       scheduler: TickWheelExecutor,
       ec: ExecutionContext,
   )(implicit F: ConcurrentEffect[F]): Client[F] =
-    makeClient(manager, responseHeaderTimeout, requestTimeout, scheduler, ec, 0)
+    new BlazeClient[F, A](manager, responseHeaderTimeout, requestTimeout, scheduler, ec, 0)
+}
+
+private class BlazeClient[F[_], A <: BlazeConnection[F]](
+    manager: ConnectionManager[F, A],
+    responseHeaderTimeout: Duration,
+    requestTimeout: Duration,
+    scheduler: TickWheelExecutor,
+    ec: ExecutionContext,
+    retries: Int,
+)(implicit F: ConcurrentEffect[F])
+    extends DefaultClient[F] {
+
+  override def run(req: Request[F]): Resource[F, Response[F]] =
+    runLoop(req, retries)
+
+  private def runLoop(req: Request[F], remaining: Int): Resource[F, Response[F]] = {
+    val key = RequestKey.fromRequest(req)
+    for {
+      requestTimeoutF <- scheduleRequestTimeout(key)
+      preparedConnection <- prepareConnection(key)
+      (conn, responseHeaderTimeoutF) = preparedConnection
+      timeout = responseHeaderTimeoutF.race(requestTimeoutF).map(_.merge)
+      responseResource <- Resource.eval(runRequest(conn, req, timeout)).recoverWith {
+        case _: SocketException if remaining > 0 && req.isIdempotent =>
+          Resource.eval(manager.invalidate(conn) *> F.pure(runLoop(req, remaining - 1)))
+      }
```

Comment:
This `recoverWith` is the meaningful part of the merge.  I didn't think this would work without a `F.continual`, as in the `Retry` middleware.  I'm concerned that we're allocating a second connection before releasing the first, but the test only allows one request per key.